### ISSUE TITLE
Highlight Color not respected on Initial Load

### ISSUE THREAD

**Issue by stephenprater** [lukas-reineke/indent-blankline.nvim#725]
### Problem

I have the Highlight group set to "NonText" on the require/setup call.

When loaded the Highlight group is NOT "NonText" - but is (presumably `IblIndent`)

Reloading the configuration resets the highlight group properly.

I am 90% certain that this is some kind of problem in my setup, but I don't understand the lifecycle well enough to debug it.

### Steps to reproduce

https://github.com/lukas-reineke/indent-blankline.nvim/assets/149870/fb19163d-f551-4404-abd3-b4ce899f3994

### Expected behavior

The initial highlight group should be what I set it to.

### Neovim version (nvim -v)

nightly

**Comment by HawkinsT** [comment]
I can confirm that this appears to be a bug and IblScope doesn't seem to be respected in any way on initialization, at least via lazy.nvim (even with `lazy = false`). A workaround for now is to set `@ibl.scope.char.1` and `@ibl.scope.underline.1` explicitly, for example:

```lua
vim.api.nvim_set_hl(0, "@ibl.scope.char.1", { fg = "#d3869b" })
vim.api.nvim_set_hl(0, "@ibl.scope.underline.1", { sp = "d3869b", underline = true })
```

Perhaps ibl setup is called too late and that's what's causing this issue?

**Comment by lukas-reineke** [comment]
Sorry I didn't see this before.

The problem is most likely that your colorscheme doesn't clear existing highlight groups.

Indent-blankline generates highlight groups if they don't exist. If ibl loads before the colorscheme it will create the highlight groups based off of the default colors. When the color scheme changes, it will generate them again, but it won't overwrite any existing groups (this is so that color schemes can set them). But if your colorscheme doesn't set them, and also doesn't clear them, you will get the default colors.

TLDR; either load ibl after you load your colorscheme, or fix the colorscheme to correctly reset existing highlight groups.

**Comment by HawkinsT** [comment]
Thanks for your response which allowed me to track down the issue to a degree. I use gruvbox.nvim, loaded via lazy.nvim with the options `lazy = false` and `priority = 1000`, so my colorscheme is loading before indent-blankline.nvim, however indent-blankline.nvim was loading before the line in my config specifying the change to `IblScope` and it seems indent-blankline.nvim never re-reads this, hence it was defaulting to linking to `LineNr` with the colour for that specified by gruvbox.nvim.

The solution is to make sure `IblScope` is specified before indent-blankline.nvim is loaded, so for example with gruvbox.nvim:
```lua
{
    "ellisonleao/gruvbox.nvim",
    lazy = false,
    priority = 1000,
    config = function()
        require("gruvbox").setup({
            overrides = {
                ["IblScope"] = { fg = "#d3869b" },
            },
        })
        vim.opt.background = "dark"
        vim.cmd.colorscheme("gruvbox")
    end,
},
```

I still have no idea how to get this to work from within the `ibl` setup function though, as the following just doesn't work for me (the scope highlights remain the same as `LineNr`):

```lua
{
    "lukas-reineke/indent-blankline.nvim",
    main = "ibl",
    config = function()
        local hooks = require("ibl.hooks")
        require("ibl").setup({
            scope = {
                highlight = "GruvboxRed" -- as an example
            },
        })
        hooks.register(hooks.type.SCOPE_HIGHLIGHT)
    end,
},
```

**Comment by lukas-reineke** [comment]
>The solution is to make sure IblScope is specified before indent-blankline.nvim is loaded

Yes, the highlight group needs to be defined before you load indent-blankline.

> I still have no idea how to get this to work from within the ibl setup function though, as the following just doesn't work for me (the scope highlights remain the same as LineNr):

This should work fine, but again, `GruvboxRed` needs to be set before you run this.

Also, this `hooks.register(hooks.type.SCOPE_HIGHLIGHT)` is wrong, `hooks.register` takes 2 arguments. What are you trying to do here?

**Comment by HawkinsT** [comment]
Sorry, I was trying to provide an example to demonstrate that it doesn't work. `GruvboxRed` is set before indent-blankline is loaded in my case, but `highlight` can just as easily be set to `ErrorMsg` or anything else that is defined at the time of neovim initializing as none work.

My config (where scope highlight doesn't work) is:

```lua
	{
		"lukas-reineke/indent-blankline.nvim", -- Indentation guides
		main = "ibl",
		config = function()
			local highlight = {
				"RainbowRed",
				"RainbowYellow",
				"RainbowBlue",
				"RainbowOrange",
				"RainbowGreen",
				"RainbowViolet",
				"RainbowCyan",
			}
			local hooks = require("ibl.hooks")

			hooks.register(hooks.type.HIGHLIGHT_SETUP, function()
				vim.api.nvim_set_hl(0, "RainbowRed", { fg = "#E06C75" })
				vim.api.nvim_set_hl(0, "RainbowYellow", { fg = "#E5C07B" })
				vim.api.nvim_set_hl(0, "RainbowBlue", { fg = "#61AFEF" })
				vim.api.nvim_set_hl(0, "RainbowOrange", { fg = "#D19A66" })
				vim.api.nvim_set_hl(0, "RainbowGreen", { fg = "#98C379" })
				vim.api.nvim_set_hl(0, "RainbowViolet", { fg = "#C678DD" })
				vim.api.nvim_set_hl(0, "RainbowCyan", { fg = "#56B6C2" })
			end)

			vim.g.rainbow_delimiters = { highlight = highlight }

			require("ibl").setup({
				scope = {
					highlight = highlight,
				},
			})

			hooks.register(hooks.type.SCOPE_HIGHLIGHT, hooks.builtin.scope_highlight_from_extmark)

			require("ibl").overwrite({
				exclude = { filetypes = { "dashboard" } },
			})
		end,
	},
```

Which is copied from this plugin's example for integration with rainbow delimiters. I tried condensing this to only that which matters but made a mistake in the hooks line in the process.

What I was trying to illustrate is that no matter what `highlight = "<some valid highlight group name>",` is set to within the ibl setup function, it doesn't appear to do anything (I've also tried this without the `hooks.register` line, following the example under documentation section `ibl.config.scope.highlight`). Can you confirm that setting the scope `highlight` variable to some value does indeed change the scope highlight for you (since this would the suggest a config issue rather than a bug with indent-blankline)?

N.B. I've just realised I've been discussing scope.highlight and the original bug report is on indent.highlight. However, to me this appears to be the same issue.

**Comment by lukas-reineke** [comment]
```lua
require("ibl").overwrite({
    exclude = { filetypes = { "dashboard" } },
})
```

This is your issue. `overwrite` will overwrite your config and set everything to the default. Anything you do before is irrelevant.

**Comment by lukas-reineke** [comment]
I made a new release to change how `overwrite` works

https://github.com/lukas-reineke/indent-blankline.nvim/pull/760

@HawkinsT for your case, I'm pretty sure that should fix it. But I'll leave the ticket open, in case it is something else for @stephenprater

**Comment by HawkinsT** [comment]
Perfect, this solved my issue. Many thanks for your assistance and quick patch, @lukas-reineke !

**Comment by stephenprater** [comment]
I changed my config stanza in Lazy to this:

```lua
	{
		"lukas-reineke/indent-blankline.nvim",
    main = "ibl",
		config = function()
      vim.api.nvim_set_hl(0, "IblIndent", { fg = "#34342d" })
      vim.api.nvim_set_hl(0, "IblWhitespace", { fg = "#34342d" })
			require("ibl").setup({
        indent = {
          char = "│",
          smart_indent_cap = true,
        },
        scope = {
          enabled = false
        }
			})
		end,
	},
```

Which if I understand correctly _should_ result in the IblIndent highlight group being set on initial load.  However - it still requires me to reload the plugin before the Highlight group is correct.  Have I missed something?

**Comment by lukas-reineke** [comment]
> Which if I understand correctly should result in the IblIndent highlight group being set on initial load. However - it still requires me to reload the plugin before the Highlight group is correct. Have I missed something?

This looks fine. But you still need to make sure that you load your colorscheme first, or you need to use `:help ibl.hooks.cb.highlight_setup()`
There is an example [in the readme](https://github.com/lukas-reineke/indent-blankline.nvim#multiple-indent-colors)

**Comment by stephenprater** [comment]
Perfection.
